### DISCUSSION TITLE
UNETR validation metric stuck on 0.5

### DISCUSSION THREAD

**Discussion by carlpe** [Project-MONAI/tutorials#1159] [Q&A] [answered]
Hi,

I have the following code for training a UNETR 3D segmentation using a single class:

```
from monai.utils import first, set_determinism
from monai.transforms import (
    AsDiscrete,
    EnsureChannelFirstd,
    Compose,
    CropForegroundd,
    LoadImaged,
    Orientationd,
    RandFlipd,
    RandCropByPosNegLabeld,
    RandShiftIntensityd,
    ScaleIntensityRanged,
    Spacingd,
    RandRotate90d,
)
from monai.losses.dice import DiceCELoss
from monai.networks.nets import UNETR
from tqdm import tqdm
from monai.networks.layers import Norm
from monai.metrics import DiceMetric
from monai.losses import DiceLoss
from monai.inferers import sliding_window_inference
from monai.data import CacheDataset, DataLoader, Dataset, decollate_batch
import torch
import matplotlib.pyplot as plt
import os
import glob
from monai.optimizers import Novograd
from monai.metrics import DiceMetric
from monai.utils import set_determinism

CUDA_LAUNCH_BLOCKING = "1"
data_root = "/home/"
path_to_target = "/home/"

train_images = sorted(glob.glob(os.path.join(path_to_target, "ct", "*.nii*")))
train_labels = sorted(glob.glob(os.path.join(path_to_target, "gt", "*.nii*")))

data_dicts = [
    {"image": image_name, "label": label_name}
    for image_name, label_name in zip(train_images, train_labels)
]
train_files, val_files = data_dicts[:90], data_dicts[90:93]
print(len(train_files))
print(len(val_files))
print("------------")
print("last img:" + train_images[-1])
print("last seg" + train_labels[-1])
print("------------")


train_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        EnsureChannelFirstd(keys=["image", "label"]),
        Orientationd(keys=["image", "label"], axcodes="RAS"),
        Spacingd(
            keys=["image", "label"],
            pixdim=(1, 1, 1),
            mode=("bilinear", "nearest"),
        ),
        CropForegroundd(keys=["image", "label"], source_key="image"),
        RandCropByPosNegLabeld(
            keys=["image", "label"],
            label_key="label",
            spatial_size=(96, 96, 96),
            pos=1,
            neg=1,
            num_samples=4,
            image_key="image",
            image_threshold=0,
        ),
        RandFlipd(
            keys=["image", "label"],
            spatial_axis=[0],
            prob=0.10,
        ),
        RandFlipd(
            keys=["image", "label"],
            spatial_axis=[1],
            prob=0.10,
        ),
        RandFlipd(
            keys=["image", "label"],
            spatial_axis=[2],
            prob=0.10,
        ),
        RandRotate90d(
            keys=["image", "label"],
            prob=0.10,
            max_k=3,
        ),
        RandShiftIntensityd(
            keys=["image"],
            offsets=0.10,
            prob=0.50,
        ),
    ]
)
val_transforms = Compose(
    [
        LoadImaged(keys=["image", "label"]),
        EnsureChannelFirstd(keys=["image", "label"]),
        Orientationd(keys=["image", "label"], axcodes="RAS"),
        Spacingd(
            keys=["image", "label"],
            pixdim=(1, 1, 1),
            mode=("bilinear", "nearest"),
        ),
        CropForegroundd(keys=["image", "label"], source_key="image"),
    ]
)

check_ds = Dataset(data=train_files, transform=train_transforms)
check_loader = DataLoader(check_ds, batch_size=1)
check_data = first(check_loader)
image, label = (check_data["image"][0][0], check_data["label"][0][0])
print(f"image shape: {image.shape}, label shape: {label.shape}")

train_ds = CacheDataset(
    data=train_files, transform=train_transforms, cache_rate=1.0, num_workers=4
)
train_loader = DataLoader(train_ds, batch_size=3, shuffle=True, num_workers=1)
val_ds = CacheDataset(
    data=val_files, transform=val_transforms, cache_rate=1.0, num_workers=0
)
val_loader = DataLoader(val_ds, batch_size=1, num_workers=0)

device = torch.device("cuda:0")

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = UNETR(
    in_channels=1,
    out_channels=2,
    img_size=(96, 96, 96),
    feature_size=16,
    hidden_size=768,
    mlp_dim=3072,
    num_heads=12,
    pos_embed="perceptron",
    norm_name="instance",
    res_block=True,
    dropout_rate=0.0,
).to(device)

loss_function = DiceCELoss(to_onehot_y=True, softmax=True)
model.load_state_dict(torch.load("mect_unetr_1class.pth"))
torch.backends.cudnn.benchmark = True
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4, weight_decay=1e-5)


def validation(epoch_iterator_val):
    model.eval()
    with torch.no_grad():
        for batch in epoch_iterator_val:
            val_inputs, val_labels = (batch["image"].cuda(), batch["label"].cuda())
            val_outputs = sliding_window_inference(val_inputs, (96, 96, 96), 4, model)
            val_labels_list = decollate_batch(val_labels)
            val_labels_convert = [
                post_label(val_label_tensor) for val_label_tensor in val_labels_list
            ]
            val_outputs_list = decollate_batch(val_outputs)
            val_output_convert = [
                post_pred(val_pred_tensor) for val_pred_tensor in val_outputs_list
            ]
            dice_metric(y_pred=val_output_convert, y=val_labels_convert)
            epoch_iterator_val.set_description(
                "Validate (%d / %d Steps)" % (global_step, 10.0)
            )
        mean_dice_val = dice_metric.aggregate().item()
        dice_metric.reset()
    return mean_dice_val


def train(global_step, train_loader, dice_val_best, global_step_best):
    model.train()
    epoch_loss = 0
    step = 0
    epoch_iterator = tqdm(
        train_loader, desc="Training (X / X Steps) (loss=X.X)", dynamic_ncols=True
    )
    for step, batch in enumerate(epoch_iterator):
        step += 1
        x, y = (batch["image"].cuda(), batch["label"].cuda())
        logit_map = model(x)
        loss = loss_function(logit_map, y)
        loss.backward()
        epoch_loss += loss.item()
        optimizer.step()
        optimizer.zero_grad()
        epoch_iterator.set_description(
            "Training (%d / %d Steps) (loss=%2.5f)"
            % (global_step, max_iterations, loss)
        )
        if (
            global_step % eval_num == 0 and global_step != 0
        ) or global_step == max_iterations:
            epoch_iterator_val = tqdm(
                val_loader, desc="Validate (X / X Steps) (dice=X.X)", dynamic_ncols=True
            )
            dice_val = validation(epoch_iterator_val)
            epoch_loss /= step
            epoch_loss_values.append(epoch_loss)
            metric_values.append(dice_val)
            if dice_val > dice_val_best:
                dice_val_best = dice_val
                global_step_best = global_step
                torch.save(
                    model.state_dict(), os.path.join(data_root, "mect_unetr_1class.pth")
                )
                print(
                    f"Model Was Saved ! Current Best Avg. Dice: {round(dice_val_best, 3)} Current Avg. Dice: {round(dice_val, 3)}"
                )
            else:
                print(
                    f"Model Was Not Saved ! Current Best Avg. Dice: {round(dice_val_best, 3)} Current Avg. Dice: {round(dice_val, 3)}"
                )

        global_step += 1
    return global_step, dice_val_best, global_step_best


max_iterations = 250000
eval_num = 1000
post_label = AsDiscrete(to_onehot=2)
post_pred = AsDiscrete(argmax=True, to_onehot=2)
dice_metric = DiceMetric(include_background=True, reduction="mean", get_not_nans=False)
global_step = 0
dice_val_best = 0.0
global_step_best = 0
epoch_loss_values = []
metric_values = []
while global_step < max_iterations:
    global_step, dice_val_best, global_step_best = train(
        global_step, train_loader, dice_val_best, global_step_best
    )
model.load_state_dict(torch.load(os.path.join(data_root, "mect_unetr_1class.pth")))

print(
    f"train completed, best_metric: {dice_val_best:.4f} "
    f"at iteration: {global_step_best}"
)

```

I was able to train it successfully (DSC=0.80) using two classes. Now I have merged all my classes into a single class and during training, the dice metric is stuck to 0.5 and it will not increase. 

<img width="385" alt="image" src="https://user-images.githubusercontent.com/54416001/211913733-ce1b21d9-53b6-4ba5-8206-ef9382f73939.png">

I suppose the issue has to do with post_pred, post_label. (onehot or something there) when going from multi-class to single-class.

Please let me know what I am doing wrong and how to fix this.

**Comment by tangy5** [comment, marked as answer]
The code looks normal, I feel the post label and pred are correct:
```
post_label = AsDiscrete(to_onehot=2)
post_pred = AsDiscrete(argmax=True, to_onehot=2)
dice_metric = DiceMetric(include_background=True, reduction="mean", get_not_nans=False)
```
I guess there is one class Dice is 1 and the other one is all 0, resulting Dice=0.5 all the time. Can you confirm the data and labels are correct after merging classes?

**Reply by carlpe** [reply]
Hi @tangy5 ,

Yes, I have checked that the two classes (artery and vein) have been merged correctly using SimpleITK.binarythresholding that looped through the ground truth-directory.


In this experiment, UNETR was working very well using two classes compared to other solutions I have tried so far. In a matter of two hours, the DSC was around 0.8 for these vessels. 

<img width="397" alt="image" src="https://user-images.githubusercontent.com/54416001/211968616-36a97463-34be-471e-a31a-85549288c707.png">


After the merge, I can now see that after we let it train for a couple of hours, it has increased slightly up from 0.5 now, but its performance is nowhere near having split the vessels into two separate classes:

<img width="405" alt="image" src="https://user-images.githubusercontent.com/54416001/211968797-b2f7539c-02b2-4359-99dd-65bb02daba1f.png">


What I have seen from other answers around the discussion forums and the issues-section are the following:

* Add softmax=True when going from one class to two in the #1006  discussion where I had some similar issue going from one class to two in some other code I tried.

* In this thread, #262 , the suggestion to use sigmoid instead of softmax is brought up. I also tried this so far, but no real difference as I could see from the beginning.


As it is no longer stuck at exactly 0.5, one possible explanation could be that it is not performing as well after all vessels are fused together into one, there might be some mistake in my dataset (I dont find this likely as I did check 5-6 of the subjects and those were in order).

**Reply by carlpe** [reply]
@KumoLiu , do you know the answer ? 😊

**Reply by KumoLiu** [reply]
Hi @carlpe, I double-checked your loss and dice function, looks correct. Just have several suggestions here.
- First of all, I think it is still worth checking the data. Visualizing data to check whether the two categories are merged correctly.
- Second, if the data-checking has no problem, perhaps you could change the sampling ratio of positive and negative samples, if the proportion of vascular foreground is relatively small, you can increase the probability.
- It is strongly recommended to change `include_background` to `False` since there is no need to count background
- A small guess here if the contrast between the artery and vein is much different it may not be a good idea to combine two categories into one, but this is just a guess as I don't know what data you are using. 

Hope it can help you, thanks!

**Reply by carlpe** [reply]
Thanks @KumoLiu 

I will try our suggestions and check the data once more. I also think that it will not be better with one merged class, but we want to try this experiment

**Reply by carlpe** [reply]
I found the error. 
Of course, a mistake in the dataset folder selection when merging of the classes together. It has been fixed now and training.

Thank you again

![image](https://user-images.githubusercontent.com/54416001/212051739-e85517ea-99f8-4752-a23b-7adfad032f4d.png)

**Reply by carlpe** [reply]
Can any of these hyper parameters be tuned in order to look for smaller structures, such as thin vessels?

```
model = UNETR(
    in_channels=1,
    out_channels=2,
    img_size=(96, 96, 96),
    feature_size=16,
    hidden_size=768,
    mlp_dim=3072,
    num_heads=12,
    pos_embed="perceptron",
    norm_name="instance",
    res_block=True,
    dropout_rate=0.0,
).to(device)
```

Glad for suggestions and insights here

**Reply by tangy5** [reply]
Hi @carlpe , if the structures are small, which means the segmentation is more challenging and harder task, you can adjust the "feature_size", "num_heads" which controls the scale of the model, for example, feature_size=32, or even 48, num_head=24, will lead to a larger models, and more powerful. But this will take more GPU memory. You could try tune these two parameters and fit for the GPU usage.

**Reply by carlpe** [reply]
Ok, I will try it 😊 thank you

**Comment by carlpe** [comment]
"It is strongly recommended to change include_background to False since there is no need to count background."  --  For this to work, do I have to change some of the parameters like to_onehot to be one less than before, since the background is not counted, or do I leave it like it is?

By the way, in this experiment, using two classes led to better results than using one class (DSC 0.83 vs 0.68). Increasing the feature_size to 32 and num_heads to 24 did produce a better result in both cases.

**Reply by KumoLiu** [reply]
Hi @carlpe, when changing `include_background` to `False`, you don't need to do anything else. It will remove the background automatically.
https://github.com/Project-MONAI/MONAI/blob/6cb4ced0ea50b427a4065b4516be61b7a073abce/monai/metrics/meandice.py#L131-L132

Thanks!

**Reply by carlpe** [reply]
Ok, thank you @KumoLiu
